### PR TITLE
Task 2 simple rest clients

### PR DESCRIPTION
**Pull request by EvgeniyTsarkov**
The following changes have been introduced:
- a simple JS api client  has been added to get the lists of categories and products
- a simple .Net api client has been added to the list of categories and products
- breadcrumps functionality fixed to get a uniform case representation

Dotnet console example
JS client example

### CODE REVIEW

File: src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+<script src="https://ajax.aspnetcdn.com/ajax/jQuery/jquery-2.0.3.min.js"></script>
+<link rel="stylesheet" href="~/css/entities-from-api.css" />
+
+<div>
+    <ul id="products" />
+</div>
+
+<script>
+    let $productsList = $('#products');
+    let apiUrl = '../api/products/';
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml:12]: Why do you specify here `../`? It should full domain URL, not "relative" path

File: src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+<script src="https://ajax.aspnetcdn.com/ajax/jQuery/jquery-2.0.3.min.js"></script>
+<link rel="stylesheet" href="~/css/entities-from-api.css" />
+
+<div>
+    <ul id="categories" />
+</div>
+
+<script>
+    let $categoriesList = $('#categories');
+    let apiUrl = '../api/categories/';
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml:12]: Why do you specify here ../? It should full domain URL, not "relative" path

File: src/NorthwindWebsite/API/Controllers/ProductsController.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ public ProductsController(IProductService productService)
     [ProducesResponseType(StatusCodes.Status200OK)]
     public async Task<ActionResult<List<Product>>> GetProducts()
     {
-        var products = await _productService.GetAll();
+        var products = await _productService.GetSimpleProducts();
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/API/Controllers/ProductsController.cs:22]: What it the the **key** difference between `GetProducts` and `GetSimpleProducts`? Looking at repository, actually, I cannot find it. Let's revert this change

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs [hunk 4]
```diff
@@ -31,19 +35,25 @@ private static List<string> ProcessSpecialCases(string path)
...
             }
 
             if (node == ImageUploadActionName)
             {
-                breadcrumbPathSegments.Add("Image Upload");
+                breadcrumbPathSegments.Add("image upload");
+                break;
+            }
+
+            if (node == FromApiJsActionName || node == FromApiDotNetActionName)
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs:54]: Let's remove this change after fixing the general comment

File: src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    let method = 'GET';
+
+    $(function(){
+        $.ajax({
+        type: method,
+        url: apiUrl,
+        success: function(products) {
+           $.each(products, function(i, product){
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml:20]: Honestly, I am not sure that JQuery `each` loop is required here since you are iterating API response, not DOM-items. Let's replace it with `for (let product of products)` loop, please

File: src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    let method = 'GET';
+
+    $(function(){
+        $.ajax({
+        type: method,
+        url: apiUrl,
+        success: function(categories) {
+           $.each(categories, function(i, category){
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml:20]: Honestly, I am not sure that JQuery `each` loop is required here since you are iterating API response, not DOM-items. Let's replace it with `for (let product of products)` loop, please

File: src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+                $categoriesList.append('<li class="list-item">' + category.categoryName + "</li>");
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Categories/FromApiJS.cshtml:21]: I think string template may be more suitable here: `<li class="list-item">'${category.categoryName }</li>`

File: src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                $productsList.append('<li class="list-item">' + product.productName + "</li>");
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Products/FromApiJS.cshtml:21]: I think string template may be more suitable here: `<li class="list-item">'${category.categoryName }</li>`

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs [hunk 9]
```diff
@@ -4,13 +4,17 @@ namespace NorthwindWebsite.Presentation.Views.Shared.Components.Breadcrumps;
 
 public class Breadcrumbs : ViewComponent
 {
-    private const string NavigationStartPoint = "Home";
-    private const string HandleActionName = "Handle";
-    private const string ImageUploadActionName = "ImageUpload";
+    private const string NavigationStartPoint = "home";
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs:7]: Let's remove this change after fixing the general comment

File: src/NorthwindWebsite/Presentation/Controllers/CategoriesController.cs [hunk 10]
```diff
@@ -60,6 +60,12 @@ public async Task<IActionResult> Images(int id)
         return File(imageByteArray, HttpContentConstants.ImageBmp);
     }
 
+    public IActionResult FromApiJs() =>
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Controllers/CategoriesController.cs:63]: Let's remove this change after fixing the general comment

File: src/NorthwindWebsite/Presentation/Controllers/ProductsController.cs [hunk 11]
```diff
@@ -79,6 +79,12 @@ public async Task<IActionResult> Delete(int id)
         return RedirectToAction("Index");
     }
 
+    public IActionResult FromApiJs() =>
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Controllers/ProductsController.cs:82]: Let's remove this change after fixing the general comment

**Review by dotnet-fizzyy — CHANGES_REQUESTED** [review]
So, regarding the task, you should not create any separate view to represent your data that arrives from API controllers, you should add just 2 clients: .Net Core console app and JS file (or it can be HTML with embedded) that should send requests to your API. In this case, you should place them in the root directory of repository, call them like (`api-clients`) and store them here. Please, let's use this approach, it will simplify your work and remove extra code.

File: src/NorthwindWebsite/API/Controllers/ProductsController.cs (hunk 3 above)
Comment (EvgeniyTsarkov) [reply src/NorthwindWebsite/API/Controllers/ProductsController.cs:22]: @dotnet-fizzyy  The difference is that in GetProducts we return a number of objects by reference. For example, in Product class we get Category by reference. As the result, it creates enormous jsons, that cannot be displayed by a browser or Postman, or processed in code. Basically what I do by returning simplified Products representation - I do not include Category or Supplier objects in them. 
For the task I need only ProductName or CategoryName. If I needed to include, for example, Supplier or Category names, I would need to create a separate dto to extract only the name from Category or Supplier and pass them to Products.

File: src/NorthwindWebsite/Configuration/MiddlewareConfiguration.cs [hunk 12]
```diff
@@ -24,6 +24,8 @@ public static class MiddlewareConfiguration
             app.UseDeveloperExceptionPage();
         }
 
+        app.UseCors("AllowCors");
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Configuration/MiddlewareConfiguration.cs:27]: Since this is a policy name, let's move it to constant and also add postfix `Policy`.
BTW, have you tried [this](https://github.com/dotnet-fizzyy/ASP.Net-Identity/blob/master/src/IdentityWebApi/Startup/Startup.cs#L84) implementation?

File: src/NorthwindWebsite/API/Controllers/ProductsController.cs (hunk 3 above)
Comment (dotnet-fizzyy) [reply src/NorthwindWebsite/API/Controllers/ProductsController.cs:22]: Ok then. So, for the future, you may resolve situation with including/excluding related entities using the following cases:
1. Pass some query param like `includeRelatives`;
2. Create a separate endpoint to return lightweight models. This option is quite rare and should have some serious changes comparing to casual endpoint;

File: api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Product.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
+﻿namespace HttpClientConsoleApp.Entities;
+
+class Product
```
Comment (dotnet-fizzyy) [inline api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Product.cs:3]: Just an advice: Don't forget about modifiers

File: api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Product.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    public short? UnitsInStock { get; set; }
+
+    public short? UnitsOnOrder { get; set; }
+
+    public short? ReorderLevel { get; set; }
+
+    public bool Discontinued { get; set; }
+
+    public virtual Category? Category { get; set; }
```
Comment (dotnet-fizzyy) [inline api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Product.cs:25]: Ain't sure you need to add `virtual` modifier here since your HTTP client is a different app and doesn't know anything about ORM used by server and also you are not using any overrides for it.
And also, since `Category` is a reference type, do you need this nullable cast?

File: api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Supplier.cs [hunk 15]
```diff
@@ -0,0 +1,5 @@
+﻿namespace HttpClientConsoleApp.Entities;
+
+public class Supplier
```
Comment (dotnet-fizzyy) [inline api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Entities/Supplier.cs:3]: Actually, do you need this class if it has no any properties? It will be omitted by your serializer if JSON-model has property, but object not

File: api-clients/JSClient.html [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    <h2> List of Categories</h2>
+    <div>
+        <ul id="categories">
+    </div>
+</body>
+</html>
+<script>
+
+    let productsApiUrl = 'https://localhost:7087/api/products';
```
Comment (dotnet-fizzyy) [inline api-clients/JSClient.html:23]: Let's convert it to `const` and extract base domain + part as separate constant to be able to easily modify it.

File: api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Program.cs [hunk 17]
```diff
@@ -0,0 +1,93 @@
+﻿using HttpClientConsoleApp.Entities;
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+
+namespace HttpClientConsoleApp;
+
+public class Program
+{
+    private const string GetProductsUrl = "https://localhost:7087/api/products";
```
Comment (dotnet-fizzyy) [inline api-clients/HttpClientConsoleApp/HttpClientConsoleApp/Program.cs:9]: Oh, also the same case with base constant as in JS file

File: src/NorthwindWebsite/Configuration/MiddlewareConfiguration.cs [hunk 18]
```diff
@@ -24,6 +24,12 @@ public static class MiddlewareConfiguration
             app.UseDeveloperExceptionPage();
         }
 
+        app.UseCors(x => x
+        .SetIsOriginAllowed(_ => true)
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Configuration/MiddlewareConfiguration.cs:28]: I advice to add tabulation here to make function chain call more obvious

**Review by dotnet-fizzyy — APPROVED** [review]
